ndonnx/_funcs.py

```diff
@@ -386,19 +388,29 @@ def equal(x, y):
 def _binary(func_name, x, y):
     x_dtype = asarray(x).dtype
     y_dtype = asarray(y).dtype
-    if (out := getattr(x_dtype._ops, func_name)(x, y)) is not NotImplemented:
-        return out
-    if (out := getattr(y_dtype._ops, func_name)(x, y)) is not NotImplemented:
-        return out
+    try:
+        if (out := getattr(x_dtype._ops, func_name)(x, y)) is not NotImplemented:
+            return out
+    except PromotionError:
+        pass
+
+    try:
+        if (out := getattr(y_dtype._ops, func_name)(x, y)) is not NotImplemented:
+            return out
+    except PromotionError:
+        pass
     raise TypeError(
```

Comment:
Would it be reasonable to simply pass the `PromotionError` through?

Reply:
I don't think so. You would end up getting different errors if you do `a + b` vs `b + a` where one is a core data type and where one is user-defined. Since we removed the explicit management of failure from the functions themselves and use the `PromotionError` to drive dispatching, you cannot raise on the first one so you shouldn't on the second either.

Reply:
But why are we not re-raising the caught `PromotionError`?

Reply:
I think it's best we speak in person/call.

Reply:
The fact that `promote` is being called is an implementation detail of the operation. A type can implement a binary operator with or without type promotion - this is also why I'm not a huge fan of driving this dispatching through exceptions, as I've mentioned elsewhere.

I think it's best we speak directly.